src/System.Management.Automation/engine/Subsystem/ISubsystem.cs

```diff
@@ -0,0 +1,60 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+
+namespace System.Management.Automation.Subsystem
+{
+    /// <summary>
+    /// Define the kinds of subsystems.
+    /// </summary>
+    public enum SubsystemKind
+    {
+        /// <summary>
+        /// Component that provides predictive suggestions to commandline input.
+        /// </summary>
+        CommandPredictor = 1,
+    }
+
+    /// <summary>
+    /// Define the base interface to implement a subsystem.
+    /// The API contracts for specific subsystems are defined within the specific interfaces/abstract classes that implements this interface.
+    /// </summary>
+    /// <remarks>
+    /// There are two purposes to have the internal member `Kind` declared in 'ISubsystem':
+    /// 1. Make the mapping from an `ISubsystem` implementation to the `SubsystemKind` easy;
+    /// 2. Make sure a user cannot directly implement 'ISubsystem', but have to derive from one of the concrete subsystem interface or abstract class.
+    /// <para/>
+    /// The internal member needs to have a default implementation defined by the specific subsystem interfaces or abstract class,
+    /// because it should be the same for a specific kind of subsystem.
+    /// </remarks>
+    public interface ISubsystem
+    {
+        /// <summary>
+        /// Gets the unique identifier for a subsystem implementation.
+        /// </summary>
+        Guid Id { get; }
+
+        /// <summary>
+        /// Gets the name of a subsystem implementation.
+        /// </summary>
+        string Name { get; }
+
+        /// <summary>
+        /// Gets the description of a subsystem implementation.
+        /// </summary>
+        string Description { get; }
+
+        /// <summary>
+        /// Gets a dictionary that contains the functions to define at the global scope.
+        /// Key: function name; Value: function script.
+        /// </summary>
+        Dictionary<string, string> FunctionsToDefine { get; }
```

Comment:
> A FunctionsToDefine requirement would make that less helpful from a user-perspective, since it would require any implementation of a given subsystem to implement the required commands

A concrete subsystem type such as `IPredictor` can have a default implementation of `FunctionsToDefine` if no functions are required for such a subsystem kind. For example, the implementation of `IPredictor` doesn't even need to know `FunctionsToDefine` exists.

But if the concrete subsystem requires certain functions to be defined, then the implementation needs to defined it, and it will be validated during registration.

> Is the idea that a PowerShell subsystem can go without an implementation, or would excluding a subsystem require a "null" implementation?

The idea is that an implementation should satisfy the requirement imposed by the concrete subsystem kind.

Reply:
> The idea is that an implementation should satisfy the requirement imposed by the concrete subsystem kind.

Yeah I understand the current idea at work, but not why that is the idea.

If PowerShell must always have the same number of subsystems supplied, how does it reduce its size or footprint?

If I don't want to use jobs and would rather not have it in my PowerShell, how do I build against the current subsystem interface to say "no job subsystem thanks"?

Reply:
> If PowerShell must always have the same number of subsystems supplied, how does it reduce its size or footprint?

I think my comment to one of your another question answers this one too. None of the subsystems would require an implementation to be registered. So, for the not-yet-existing `System.Management.Automation.Core`, it should work with local in-proc powershell language processing without having any subsystem registered. You will only register the subsystems you need, and leave the others unregistered.

Reply:
Yes, absolutely -- answered in the other comment